### ISSUE TITLE
Is it better to run format before lint?

### ISSUE THREAD

**Issue by jamesmyatt** [astral-sh/ruff-pre-commit#74]
There are various ruff lint rules that are automatically fixed by ruff format, such as extra whitespace (W291) or semicolons (E703). Hence, is it better to suggest running the formatting hook before the linting hook, when using both?

**Comment by woctezuma** [comment]
> When running with `--fix`, Ruff's lint hook should be placed _before_ Ruff's formatter hook, and
> _before_ Black, isort, and other formatting tools, as Ruff's fix behavior can output code changes
> that require reformatting.
> 
> When running without `--fix`, Ruff's formatter hook can be placed before or after Ruff's lint hook.
> 
> (As long as your Ruff configuration avoids any [linter-formatter incompatibilities](https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules),
> `ruff format` should never introduce new lint errors, so it's safe to run Ruff's format hook _after_
> `ruff check --fix`.)

**Comment by jamesmyatt** [comment]
Thanks @woctezuma . I suppose my question is whether in the "without `--fix`" case, it's better to recommend that format be run first, because some of the formatting fixes will avoid linter errors, rather than to just say it "can be placed before or after".

**Comment by MichaReiser** [comment]
@jamesmyatt In the case where you don't use `--fix` I think it does make sense to run `format` before lint if you have some `pycodestyle` rules enabled. However, our general recommendation is to disable any formatting related lint rules if you use the formatter because they don't need to run, the formatter guarantees consistent formatting.

**Comment by jamesmyatt** [comment]
Thanks @MichaReiser . That makes sense. Is there a list of those?

**Comment by MichaReiser** [comment]
I don't think so. We should probably add one. We plan to have an entire category for formatting related lint rules which should make this easier in the future. 

For now, it's mainly the pycodestyle rules (`E1`, `E2`, and `E3`)

**Comment by Mathieson** [comment]
I'm finding the linter reports incorrect lines for issues when run before the formatter, which makes sense considering the formatter might break up a single line into multiple, changing the composition of the file. The incorrect lines adds extra effort when tracking down the reported issues to fix them.

The line numbers are correct when the linter runs after the formatter.
